packages/gatsby/src/utils/webpack-utils.ts

```diff
@@ -195,12 +195,22 @@ export const createWebpackUtils = (
     },
 
     miniCssExtract: (options = {}) => {
-      return {
-        options,
-        // use MiniCssExtractPlugin only on production builds
-        loader: PRODUCTION
+      let loader: any
+      let hmr = false
+      if (process.env.GATSBY_EXPERIMENTAL_DEV_SSR) {
+        loader = MiniCssExtractPlugin.loader
+        if (stage === `develop`) {
+          hmr = true
```

Comment:
the heart of the change — if the stage is `develop`, just set hmr to `true` for mini-css-extract-plugin. We'll keep this behind a flag w/ DEV_SSR for now.